man/yaws.1
.TH YAWS "1" "" "" "User Commands" -*- nroff -*-
.SH NAME
yaws \- yet another webserver
.SH SYNOPSIS
.B yaws
[\fIOPTIONS\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Yaws is fast lightweight webserver. It can run as daemon or
in interactive mode where it is possible to directly interact
with the webserver. Yaws is particularly good at generating
dynamic content. See the user docs for more information on that topic.

.SH DAEMON/SERVER options


.TP
\fB\-i | --interactive\fR
Interactive mode. This will start yaws in interactive mode with an erlang
prompt. All error_logger messages will be written to the tty as well in this
mode. Use this when developing yaws code.
.TP
\fB\-w | --winteractive\fR
Cygwin inteactive mode (werl)
.TP
\fB\-D | --daemon\fR
Daemon mode. This will start yaws as a daemon.
.TP
\fB\--heart\fR
This will cause the yaws system to be automatically restarted
in case it should crash. This switch also requires the \fI--daemon\fR
switch to be present.
.TP
\fB\--heart-restart=C,T\fR
This controls the number of restarts in a given time period that heart
tolerates before refusing to restart Yaws. By default, heart allows up
to 5 restarts within a 60 second period before refusing to restart
Yaws again. This option allows up to \fIC\fR restarts in \fIT\fR
seconds instead. To allow infinite restarts, set both \fIC\fR and
\fIT\fR to 0. This switch automatically enables the \fI--heart\fR
switch.
.TP
\fB\--debug\fR
Debug mode. This will produce some auxiliary error output for some
error conditions. It will also start the otp sasl lib for additional
error printouts.
.TP
\fB\--nodebug\fR
Non-debug mode. This is useful for running interactively via the
\FI-i\fR option but without incurring the performance penalties of
debug mode.

.TP
\fB\--conf file\fR
Use a different configuration file than the default. If the configuration
parameter \fIconfig\fR is set, yaws use it as default configuration file. Else,
The default configuration file when running as root is /etc/yaws/yaws.conf. When
running as a non priviliged user, yaws will search for its configuration file in
the following order. First in $HOME/yaws.conf, then in ./yaws.conf and finally
in /etc/yaws/yaws.conf.
.TP
\fB\--runmod module\fR
Tells yaws to call \fImodule:start/0\fR at startup. This makes it possible
to startup user specific applications together with yaws.

.TP
\fB\--pa path\fR
Add path to the yaws system search path

.TP
\fB\--tracetraf\fR
Traffic trace mode. All traffic will be written to a trace file called trace.traffic
in the log directory.

.TP
\fB\--tracehttp\fR
HTTP trace mode. All HTTP messages will be written to a trace file called trace.http
in the log directory.

.TP
\fB\--traceout\fR
When yaws is put into trace mode using either --tracetraf or --tracehttp, traces are written to files. If we provide the --traceout flag, the trace will also be written to stdout.

.TP
\fB\--trace\fR
Sames as --tracetraf --traceout. I.e. trace everything and write to stdout.

.TP
\fB\--mnesiadir dir\fR
Start Mnesia in directory <dir>

.TP
\fB\--sname xxx\fR
Start yaws as a distributed erlang node with name <xxx>
using the unqualified hostname as nodename postfix

.TP
\fB\--disable-kpoll\fR
By default, yaws starts erlang with \fI+K true\fR. This flag
reverses that.

.TP
\fB\--name xxx\fR
Start yaws as a distributed erlang node with name <xxx> using the
fully qualified hostname as nodename postfix

.TP
\fB\--proto_dist Mod\fR
Use module Mod for erlang distribution. This is typically only used
when we want to run erlang distribution over SSL.

.TP
\fB\--erlarg STRING\fR
Pass STRING as an additional argument to the "erl" program. If STRING
comprises multiple words, you must quote it so that your shell passes
it to yaws as a single argument. If STRING contains any single quote
characters, you must quote each of them as well. For example, to pass
the option \fI-env NAME O'Keeffe\fR to "erl" from a Bourne-compatible
shell:
.PP
.nf
.ce
--erlarg "-env NAME O\\'Keeffe"
.fi

.TP
\fB\--id ID\fR
This flag sets the id. If we're starting a daemon (or an interactive
system) it gives the Yaws server the identity ID. This means that the
server will write all internal files into the directory
$HOME/.yaws/yaws/ID.

Yaws also creates a file called
$HOME/.yaws/yaws/ID/CTL which contains the portnumber the daemon
is listening on for control request by the control command such
as "yaws --hup" etc.

If we're invoking a control command which should perform some
control function on the daemon, we may have to give the --id flag also
to the control command. If we don't do this the control command
may interact with the wrong daemon due to finding the wrong "ctl" file.

The daemon may also optionally specify the "id" in the yaws.conf
configuration file.

.TP
\fB\--umask MASK\fR
Set the umask for the daemon to MASK.

.TP
\fB\--encoding latin|unicode\fR
Set the config file encoding (default: latin1).

.SH CONTROL OPTIONS
.PP
The following list of options are are used to control the daemon
from the "outside" while it is running.

.TP
\fB\--hup [--id ID]\fR
HUP the daemon. This forces the daemon to reread the configuration file.
It also makes
the daemon empty all its internal content caches.
Hence when updating the doc root,
HUPing the daemon is the fastest way to see the content updates.
.TP
\fB\--stop [--id id]\fR
Stop the daemon (called id)
.TP
\fB\--ls \fR
Lists current ids and status of all yaws servers on localhost. In practice this
amounts to a listdir in $HOME/.yaws/yaws - and check whether the different
systems who has created files there are alive.
.TP
\fB--status [--id id]\fR
Query a running yaws daemon for its status, and print it.
.TP
\fB--stats [--id id]\fR
Query a running yaws daemon for its statistics, and print it.
.TP
\fB--running-config [--id id]\fR
Query a running yaws daemon for its current configuration, and print it.
This can be useful when attempting to figure out how to set config
in embedded mode. Configure yaws to you liking in non-embedded mode, run
this command and use the output to populate the embedded mode records.
.TP
\fB--load Modules [--id id]\fR
Try to (re)load erlang modules into a running daemon.  This is useful
after modifying appmods or modules used by scripts.

.TP
\fB--debug-dump  [--id id]\fR
Produce a debug dump on stdout. In particular this code lists what
we refer to as suspicious processes. I.e. processes that might be
hanging or processes that are "large" - hardcoded to 40k words.

.TP
\fB--ctltrace [--id ID] http | traffic | off\fR
Control the trace capabilities of a running yaws daemon. If the
http or traffic option is given, the daemon will write a log
for debug purposes into the logdir.

.TP
\fB--wait-started[=T] [--id ID] \fR
Waits at most 30 seconds for the server to start. Exits with 0 if
server is running, 1 otherwise. Typically useful in test scripts. The
default 30 seconds can be modified by appending \fI=T\fR to the option,
where \fIT\fR is the desired number of seconds to wait for the server
to start.

.SH MISC OPTIONS

.TP
\fB--check YawsFile [IncDirs ....]\fR
Test compile a `.yaws' file. Useful in Makefiles when we want to ensure
that all .yaws files are syntactically correct
.TP
\fB\--version\fR
output version information and exit

.SH ENVIRONMENT VARIABLES
.TP
\fBHOME\fR
Is used to determine where we write the temporary files. By default
all tmp files end up in $HOME/.yaws. This includes the JIT files
that are the result of processed .yaws files and also the so called
control file that is used by the daemon to write the port number
to which it is listening for control commands such as "yaws --status"

Thus HOME is the handle we use in the control commands to
find the control file so that we know where to connect to.

.TP
\fBYAWSHOME\fR
Can be used to override the HOME variable. This is useful when we
for example are running yaws under port binding programs such as
authpriv.

It's useful by distros that don't want Yaws to write any files
ever in the HOME directory of root.



.SH AUTHOR
Written by Claes Wikstrom
.SH "SEE ALSO"
.BR yaws.conf (5)
.BR erl (1)


www/news
Wed Sep 30 09:26:55 CEST 2015 Yaws 2.0.1
OTP 18.1 broke the logging code in Yaws (capflam)
use text/xml as SOAP MIME type (steve)
travis is now working for Yaws
cache refres bugs fixed (capflam)
max_connection bugs fixed (steve, andrei)
ssl client_renegotiation option supported

Mon Jul 13 19:27:06 CEST 2015 Yaws 2.0
Fix Yaws build for Erlang 18.0 (now() deprecated)
Renew self-signed ssl certificat to use the right signature algorithm
support for the SSL 'log_alert' option
expires fixes
Fix possible XSS in yaws_outmod:crashmsg/3
rebar work.
Rework how SSL protocol_version is handled
Many minor bugfixes, tests and also docs.
 
Tue Nov 11 17:17:37 CET 2014 Yaws 1.99
Major release, many enhancenements and bugfixes. Too many to list actually, see the git logs. The major ones though:
combine multiple Accept-Encoding headers (steve)
Massive autotools refactoring & cleanup (Christopher Faulet)
Add SSL protocol_version configuration element (steve)
Add 'honor_cipher_order' option in ssl part (Christopher Faulet)

Mon Nov  4 19:22:42 CET 2013 Yaws 1.98
Bad HTTP 3.0 message sent to yaws, crashed the server entirely. Note this is crucial for yaws servers deployed on internet. (klacke)
Use Yaws linked-in driver by default to send large file (capflam)
Rebar work (Steve, Joseph Wayne Norton)
Wed Oct 16 05:40:25 CEST 2013 Yaws 1.97
Webdav work, all webdav is now an appmod (tjeerd)
New serveralias directive.  Now it is possible to set alternate names for a virtual host. A server alias  may contain wildcards. Inside a server definition wecan now have e.g serveralias = *.server.domain.com *.server?.domain.com (capflam)
pam bug found by Jari Hellman
optimized ssi code by storing binaries as the data chunks instead of lists of chars (klacke)
Support message body in DELETE requests  (capflam)
Group automatically the server configurations in embedded mode. Now, Yaws groups the server configurations by itself when it is started    in embedded mode. This is done when yaws_api:setconf/2 is called. (capflam)
Use crypto:hash instead of crypto:sha to be compatible with R16B1. (Karolis Petrauskas )
Use proper linker command for drivers on OpenBSD (Francis Joanis )
Try to make the log directory if it doesn't exist. This happens only if logdir has a plus sign in front of it. E.g         logdir = +path/to/log. (capflam, James Lee)
Incorrect X-Forwarded-For header (Manuel Durán Aguete)
Redhat /etc script bug fix by flavio-fernans
Install procedure for FreeBSD wasn't correct (klacke)
More rebar support (Steve, Tuncer)
Fix option to enable/disable sendfile support in the configure script. file:sendfile/5 implementation is buggy (in R15 & R16). When async-threads are enabled, in efile driver, the tcp socket is set in blocking mode and thesendfile syscall is executed on an async-thread. So an unresponsive client could block it for a very long time and therefore block the async-thread. In this way, all async-threads could be easily blocked. So, by default, the use of sendfile is disable.  (capflam)
Add support of many options for tcp/ssl sockets (capflam)
Fix mime_types.erl generation for a local installation of Yaws,  (capflam)

Tue Mar 19 17:53:00 CET 2013 Yaws 1.96
Websocket release, lots of work and rewrites of the websocket code which is now topnotch.
Handle relative paths when Yaws is started in embedded mode  (capflam)
ensure compatibility with Erlang R13B04 (steve)
Handle case with DOS attack towards pam login with embedded NUL chars in username or password. The protocol cannot handle NUL chars and Yaws dies (klacke)
Extend syntax of redirect block to allow an optional status code. (capflam)
Support multipart/form-data "name" with backslash at the end. (Johannes Weißl)
Make be possible to chain appmods (capflam)
Flush remaining data in case of 3xx redirect (capflam)
Add support of the 'Vary' header in response (capflam)
Add access functions for #gconf{} and #sconf{} records (capflam)
Refactor WebSockets and add support of optional callback functions (capflam)

Wed Oct 17 19:57:05 CEST 2012 Yaws 1.95
Feature relase, 
The --heart option didn't work as expected when yaws ran out of filedescriptors - capflam
Major webdav overhaul by Tjeerd van der Laan. Lot's of improvements and general bugfixes.
Added sorting of subconfigdir - Tjeerd van der Laan
Added a fast dispatch module. This is for applications that only want the HTTP portion of yaws, i.e applications with very high performance requirements. A lot of the normal book keeping, tracing etc is ignored.
Added RFC6265 compliant variant of yaws_api:set_cookie/3 - serge
Fix SSL socket wrapping - capflam/Jean-Charles Campagne
Parse extensions and trailing headers of chunked transfer encoded requests - capflam
Optimize the ACL checks during the request's authentication -capflam
Added support to configure ciphers in ssl - klacke
dialyzer cleanup/improvements - Kenji Rikitake
makefile / mime_type_c.beam cleanup - Kenji Rikitake
add startup script for FreeBSD 9 (Thomas Elsgaard)
make sure "rebar eunit" passes - steve
new yaws_api:reformat_header/2 function - steve
support Raspberry Pi in configure.in (Anders Nygren)
Add basic websocket test - JD Bothma
Make the mime types mappings configurable - capflam
Add index_files directive into the server part configuration - capflam
Refactor Set-Cookie/Cookie header parsing to follow RFC6265 - capflam
enabled http client library options customization - Kaloyan Dimitrov

Sun Jun 24 23:47:57 CEST 2012 Yaws 1.94
Bugfix release for bugs that sneaked into 1.93
The random patch for 1.93 wasn't good enough as discovered by Sergei Golovan, we need to cater for non printable chars (Sergei Golovan)
add support for W3C Server-Sent Events (Steve)
add easy getter/setter functions for #headers records (Steve)
add reverse proxy intercept module capability (Steve)
Patch from Nico Kruber to fix compile issue on OTP R13

Wed Jun 20 20:22:11 CEST 2012 Yaws 1.93
Security release

win32/yaws.c
/* win32 version of yaws starup script */
/* Author: klacke@hyber.org (Jan 2009) */


#include <windows.h>
#include <stdio.h>

#define BSIZ 16000

static unsigned char path[BSIZ];
static unsigned char *fpath;

static void readreg() {
    HKEY hKey;
    DWORD bsz = BSIZ;
    // Check where yaws is installed, we need to -pa that path
    if (RegOpenKeyEx(HKEY_LOCAL_MACHINE,
                     "SOFTWARE\\Hyber\\Yaws",
                     0, KEY_READ, &hKey) != ERROR_SUCCESS)
        {
            printf ("Cannot find yaws installation in the registry\n");
            exit(1);
        }
    memset(path, 0, BSIZ);
    /*int ret = */RegQueryValueEx(hKey, "DIR", NULL, NULL, path, &bsz);
    fpath = (unsigned char*)malloc(strlen((char*)path));
    int i;
    for(i = 0; i<strlen((char*)path); i++ ) {
        if (path[i] == '\\') {
            fpath[i] = '/';
        }
        else
            fpath[i] = path[i];
    }
    fpath[i] = 0;

}


int charg(char **argv, char *s1, char *s2, int p) {
    if (strcmp(s1, argv[p]) == 0) return 1;
    if ((s2 != NULL) && strcmp(s2, argv[p]) == 0) return 1;
    return 0;
}


int runwait(char *execString) {

    STARTUPINFO          si = { sizeof(si) };
    PROCESS_INFORMATION  pi;

    //printf("execString: %s \n", execString);

    if(CreateProcess(0, execString, NULL, NULL, FALSE, 0, 0, 0, &si, &pi))
        {
            unsigned long ret = 0;

            // wait for process to finish
            WaitForSingleObject(pi.hProcess, INFINITE);
            if (GetExitCodeProcess(pi.hProcess, &ret) == 0)
                ret = 1;
            CloseHandle(pi.hProcess);
            CloseHandle(pi.hThread);
            return ret;
        }
    printf("Failed to create the process entirely\n");
    printf("Tried to invoke: <%s> \n", execString);
    printf("Make sure you have <erl> in your environment PATH \n");
    return 1;
}



int nosh(char *s) {
    char buf[BUFSIZ];
    sprintf (buf, "erl -noshell -pa \"%s/ebin\" %s ", fpath, s);
    return runwait(buf);
}

int help() {
    printf("usage:  "
""
"       yaws -i | --interactive       -- interactive (no daemon) mode"
"       yaws -w | --winteractive      -- interactive (werl) "
"       yaws --daemon                 -- daemon mode"
""
""
""
"       Auxiliary flags for the daemon: "
"            --id Id          --  set system id"
"            --debug          --  debug mode "
"            --conf File      --  set config file"
"            --tracetraf      --  trace traffic"
"            --tracehttp      --  trace http traffic"
"            --traceout       --  trace output to stdout"
"            --version        --  print version"
"            --pa path        --  add load path"
"            --mnesiadir dir  --  start Mnesia in dir"
"            --sname xxx      --  start with sname xxx"
"            --name xxx       --  start with name xxx"
"            --runmod mod     --  call mod:start/0 at startup"
"            --erlarg X       --  pass argument X to $erl"
""
"ctl functions ... "
"        yaws --hup [--id ID]                 -- hup the daemon, reload conf"
"        yaws --stop [--id ID]                -- stop the daemon "
"        yaws --debug-dump [--id ID]          -- produce a debug dump "
"        yaws --status [--id ID]              -- query the daemon status "
"        yaws --load Modules                  -- load modules "
"        yaws --ls                            -- list Yaws nodes and their status"
"        yaws --ctltrace traffic|http         -- toggle trace of running daemon");

    return(0);
}



int main(int argc, char**argv) {

    int interactive  = 1;
    int winteractive  = 0;
    int daemon = 0;
    char *conf = NULL;
    char *sname = NULL;
    char *name = NULL;
    int debug = 0;
    char *runmod = NULL;
    char *trace = NULL;
    char *traceoutput = NULL;
    char tbuf[BUFSIZ];
    char paBuf[BUFSIZ];
    char execString[BSIZ];
    char mnesia[255];
    char erlarg[255];
    char *id = "default";

    memset(paBuf, 0, BUFSIZ);
    memset(tbuf, 0, BUFSIZ);
    memset(execString, 0, BSIZ);

    readreg();
    sprintf(paBuf, " -pa \"%s/ebin\" ", fpath);

    int p = 1;
    while (p < argc) {
        if (charg(argv, "-i", "--interactive", p))   {
            interactive = 1;
        }
        else if (charg(argv, "-w", "--winteractive", p)) {
            winteractive = 1;
        }
        else if (charg(argv, "-D", "--daemon", p)) {
            interactive = winteractive = 0;
            daemon = 1;
        }
        else if (charg(argv, "--debug", NULL, p)) {
            debug = 1;
        }
        else if (charg(argv, "-t", "--tracetraf", p)) {
             trace=" -yaws trace traffic ";
        }
        else if (charg(argv, "-T", "--tracehttp", p)) {
             trace=" -yaws trace http ";
        }
        else if (charg(argv, "-x", "--traceout", p)) {
            traceoutput = " -yaws traceoutput ";
        }
        else if (charg(argv, "--trace", NULL, p)) {
            traceoutput = " -yaws traceoutput ";
            trace = " -yaws trace traffic ";
        }
        else if (charg(argv, "--mnesiadir", "-M", p)) {
            sprintf(mnesia, "-mnesia dir \"%s\" -run mnesia start ",argv[++p]);
        }
        else if (charg(argv, "-c", "--conf", p))  {
            conf = argv[++p];
        }
        else if (charg(argv, "--id", NULL, p))  {
            id = argv[++p];
        }
        else if (charg(argv, "-pa", NULL, p))  {
            sprintf(tbuf, " -pa \"%s\" ", argv[++p]);
            strcat(paBuf, tbuf);
        }
        else if (charg(argv, "--runmod", "-r", p)) {
            runmod = argv[++p];
        }
        else if (charg(argv, "--hup", "-h", p)) {
            return nosh("-s yaws_ctl hup default");
        }
        else if (charg(argv, "--stop", "-s", p)) {
            return nosh("-s yaws_ctl stop default");
        }
        else if (charg(argv, "--status", "-S", p)) {
            return nosh("-s yaws_ctl status default");
        }
        else if (charg(argv, "--debug-dump", NULL, p)) {
            return nosh("-s yaws_ctl debug_dump default ");
        }
        else   if (charg(argv, "--version", "-v", p)) {
            nosh("-s yaws printversion");
            exit(0);
        }
        else   if (charg(argv, "--sname", "-sname", p)) {
             sname = argv[++p];
        }
        else   if (charg(argv, "--name", "-name", p)) {
             name = argv[++p];
        }
        else   if (charg(argv, "--erlarg", "-erlarg", p)) {
            sprintf(tbuf, " %s ", argv[++p]);
            strcat(erlarg, tbuf);
        }
        else {
            return help();
        }
        p++;
    }

    if (winteractive)
        sprintf(execString, "werl.exe  %s %s ", erlarg, paBuf);
    else
        sprintf(execString, "erl.exe  %s %s ", erlarg, paBuf);

    if (debug)
        strcat(execString, " -boot start_sasl -yaws debug ");
    if ((interactive == 0 && winteractive == 0) || daemon == 1)
        strcat(execString, " -detached ");

    if (conf == NULL)
        sprintf(tbuf, " -conf \"%s\\yaws.conf\" ", path);
    else
        sprintf(tbuf, " -conf \"%s\" ", conf);
    strcat(execString, tbuf);

    sprintf(tbuf, " -run yaws -yaws id %s ", id);
    strcat(execString, tbuf);

    if (mnesia != NULL)
        strcat(execString, mnesia);
    if (trace != NULL)
        strcat(execString, trace);
    if (traceoutput != NULL)
        strcat(execString, traceoutput);

    if (sname != NULL) {
        sprintf(tbuf," -sname %s ", sname);
        strcat(execString, tbuf);
    } else if (name != NULL) {
        sprintf(tbuf, " -name %s ", name);
        strcat(execString, tbuf);
    }
    if (runmod != NULL) {
        sprintf(tbuf, " -runmod %s ", runmod);
        strcat(execString, tbuf);
    }

    return runwait(execString);
}



